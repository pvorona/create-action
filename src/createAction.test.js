const createAction = require('./createAction')
const PropTypes = require('prop-types')

it('Creates action object', () => {
    const fetchStart = createAction('FETCH_START', 'id', 'type')
    expect(fetchStart('1', 'cat')).toMatchSnapshot()
})

it('Invalid number of parameters throws Error', () => {
    const fetchStart = createAction('FETCH_START', 'id', 'type')
    expect(fetchStart).toThrow(Error)
})

it('toString returns action type', () => {
    const creator = createAction('INIT_APPLICATION')
    expect(creator.toString()).toEqual('INIT_APPLICATION')
})

it('Accepts payload creator function', () => {
    const setSidebarExpanded = createAction(
        'SET_SIDEBAR_EXPANDED',
        (expanded = false, id) => ({ expanded, id }),
    )
    expect(setSidebarExpanded()).toMatchSnapshot()
})

it('Provides one unnamed argument as payload object', () => {
    const fetchStart = createAction('FETCH_START')
    expect(fetchStart({ id: '2', schema: 'TOY' })).toMatchSnapshot()
})

it('Provides all unnamed arguments as payload array', () => {
    const fetchStart = createAction('FETCH_START')
    expect(fetchStart('2', 'TOY')).toMatchSnapshot()
})

describe('Accepts configuration object', () => {
    it('Accepts argument names as payload property', () => {
        const fetchStart = createAction({
            type: 'FETCH_START',
            payload: ['id', 'schema'],
        })
        // const fetchStart = (id = '0', schema) => ({
            // type: 'FETCH_START',
            // id,
            // schema,
        // })
        expect(fetchStart('guest', 'cat')).toMatchSnapshot()
    })

    it('Throws Error with invalid number of parameters', () => {
        const fetchStart = createAction({
            type: 'FETCH_START',
            payload: ['id', 'schema'],
        })
        expect(() => fetchStart('guest')).toThrow(Error)
    })

    it('toString returns action type', () => {
        const creator = createAction('INIT_APPLICATION')
        expect(creator.toString()).toEqual('INIT_APPLICATION')
    })

    it('Accepts payload creator as payload property', () => {
        const fetchStart = createAction({
            type: 'FETCH_START',
            payload: (id, schema = SCHEMAS.DEFAULT_SCHEMA) => ({ id, schema })
        })
        expect(fetchStart('guest', 'cat')).toMatchSnapshot()
    })

    it('Validates arguments by using PropTypes', () => {
        const fetchStart = createAction({
            type: 'FETCH_START',
            id: PropTypes.string.isRequired,
            schema: PropTypes.string.isRequired,
        })
        fetchStart(1,2)
    })
})



// arguments | config
//




// export const fetchStart = createAction({
//     type: 'FETCH_START',
//     payloadCreator: (id, schema = SCHEMAS.DEFAULT_SCHEMA) => ({ id, schema })
// })
// export const fetchStart = createAction({
//     type: 'FETCH_START',
//     id: PropTypes.string.isRequired,
//     schema: PropTypes.string.isRequired,
// })

// export const fetchStart = createAction({
//     type: 'FETCH_START',
//     payload: ['id', 'schema'],
// })
// export const fetchStart = createAction('FETCH_START', {
//     id: PropTypes.string.isRequired,
//     schema: PropTypes.string.isRequired,
// })



// export const fetchStart = createAction('FETCH_START', (id, schema = SCHEMAS.DEFAULT_SCHEMA) => ({ id, schema }))
// export const fetchStart = createAction('FETCH_START', 'id', 'schema')

// export const fetchStart = createAction()
