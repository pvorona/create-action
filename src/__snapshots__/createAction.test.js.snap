// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accepts configuration object Accepts argument names as payload property 1`] = `
Object {
  "id": "guest",
  "schema": "cat",
  "type": "FETCH_START",
}
`;

exports[`Accepts configuration object Accepts payload creator as payload property 1`] = `
Object {
  "id": "guest",
  "schema": "cat",
  "type": "FETCH_START",
}
`;

exports[`Accepts payload creator function 1`] = `
Object {
  "expanded": false,
  "id": undefined,
  "type": "SET_SIDEBAR_EXPANDED",
}
`;

exports[`Creates action object 1`] = `
Object {
  "id": "1",
  "type": "cat",
}
`;

exports[`Provides all unnamed arguments as payload array 1`] = `
Object {
  "payload": Array [
    "2",
    "TOY",
  ],
  "type": "FETCH_START",
}
`;

exports[`Provides one unnamed argument as payload object 1`] = `
Object {
  "payload": Object {
    "id": "2",
    "schema": "TOY",
  },
  "type": "FETCH_START",
}
`;
